### PR TITLE
`Development`: Cleanup all migrations

### PR DESCRIPTION
**Pull request by krusche**
# DO NOT DEPLOY, THIS PR CONTAINS DATABASE MIGRATIONS WHICH CANNOT BE REVERTED

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

### Motivation and Context
Keep the migration history as short as possible

### Steps for Testing

Test the whole application and make sure nothing is broken

#### Exam Mode Testing

Test the whole application and make sure nothing is broken

#### Code Review
- [ ] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2
#### Exam Mode Test
- [ ] Test 1
- [ ] Test 2

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+     * If an SQLException occurs during the update process, a RuntimeException is thrown, indicating
+     * that the checksum update operation has failed. This failure needs to be addressed to ensure
+     * the integrity and consistency of the database schema migration process.
+     *
+     * @param newVersion The new version of the application for which the initial schema checksum needs to be updated.
+     * @throws RuntimeException If updating the checksum fails due to an SQLException, encapsulating the original exception.
+     */
+    private void updateInitialChecksum(String newVersion) {
+        // SQL statement with a placeholder for the newVersion parameter
+        String updateSqlStatement = """
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java:205]: What about the rest of the table? Shouldn't we clear everything but the Initial schema generation first?

File: src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+ * <p>
+ * Steps for new database migration paths:
+ * <ol>
+ * <li>Re-create the new initial scheme from the database after applying all existing liquibase migrations.</li>
+ * <li>Delete all existing liquibase migrations except the new initial scheme and *_cleanup.xml (adapt *_cleanup.xml with the latest date).</li>
+ * <li>Delete all Java migrations that have been executed before.</li>
+ * <li>Check that the new initial scheme is compatible with MySQL and Postgres and can be started from scratch in both environments.</li>
+ * <li>Add the new migration path in the constructor of DatabaseMigration.</li>
+ * <li>Verify that the migration works with the designated path (ideally with a local database and with a dump from a test / production system).</li>
+ * <li>Document the migration carefully in the release notes and the online documentation.</li>
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java:79]:
```suggestion
 * <li>Apply all existing liquibase migrations.</li>
 * <li>Re-create the new initial scheme from the database.</li>
 * <li>Delete all existing liquibase migrations except the new initial scheme and *_cleanup.xml</li>
 * <li>Rename *_cleanup.xml to match the latest date.</li>
 * <li>Delete all Java migrations that have been executed before.</li>
 * <li>Check that the new initial scheme is compatible with MySQL and Postgres and can be started from scratch in both environments.</li>
 * <li>Add the new migration path in the constructor of DatabaseMigration.</li>
 * <li>Verify that the migration works with the designated path (ideally with a local database and with a dump from a test / production system).</li>
 * <li>Document the migration carefully in the release notes and the online documentation.</li>
```
Questions that I recommend answering in that section:

1 How does on apply the existing liquibase migrations? Are you talking about starting a running application? Starting a new application? Executing a command?
2 How? => Gradle & the command
3 Where does the *_cleanup.xml come from? Auto-Generated? If it's not, why can the content stay as it is?
5 Why "that have been executed before"? After running Liquibase (unless via command) I would expect that all migrations have been run.
7 How do I do that? What version do I select? Looking at the versions above, it's the last version before the new release?
8 What does "verify" mean? No errors on start-up? Should one go through all of the schema? Or just make sure that basically nothing changed, including data?

File: src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    public String getPreviousVersionString() {
+        return previousVersionString;
+    }
+
+    /**
+     * Checks against the defined migration paths to determine if the current version of the database
+     * is compatible or requires migration. This method ensures that the database schema is prepared
+     * and up-to-date before proceeding with the application startup.
+     */
+    public void checkMigrationPath() {
```
Comment (julian-christl) [inline src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java:114]: Feels like this method could be static and take the current version and datasource.
The migration paths could just be a static final list.

Then we could simply do the check in the `LiquibaseConfiguration`: `DatabaseMigration.checkMigrationPath(currentVersion, datasource);`

What do you think?

File: src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+     * @throws RuntimeException If updating the checksum fails due to an SQLException, encapsulating the original exception.
+     */
+    private void updateInitialChecksum(String newVersion) {
+        // SQL statement with a placeholder for the newVersion parameter
+        String updateSqlStatement = """
+                UPDATE DATABASECHANGELOG
+                SET MD5SUM = null,
+                    DATEEXECUTED = now(),
+                    DESCRIPTION = 'Initial schema generation for version ' || ?,
+                    LIQUIBASE = '4.27.0',
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java:210]: not an easy one I guess

File: src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java (hunk 1 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/config/migration/DatabaseMigration.java:205]: No, this is done within a liquibase migration

**Pull request merged**